pkg/receive/handler_test.go

```diff
@@ -94,47 +87,137 @@ func TestCountCause(t *testing.T) {
 			}),
-			f:   isConflict,
-			out: 3,
+			threshold: 1,
+			exp:       errConflict,
+		},
+		{
+			name: "matching multierror many, one above threshold",
+			err: terrors.MultiError([]error{
+				storage.ErrOutOfOrderSample,
+				errConflict,
+				tsdb.ErrNotReady,
+				tsdb.ErrNotReady,
+				tsdb.ErrNotReady,
+				errors.New("foo"),
+			}),
+			threshold: 2,
+			exp:       tsdb.ErrNotReady,
+		},
+		{
+			name: "matching multierror many, both above threshold, conflict have precedence",
+			err: terrors.MultiError([]error{
+				storage.ErrOutOfOrderSample,
+				errConflict,
+				tsdb.ErrNotReady,
+				tsdb.ErrNotReady,
+				tsdb.ErrNotReady,
+				status.Error(codes.AlreadyExists, "conflict"),
+				errors.New("foo"),
+			}),
+			threshold: 2,
+			exp:       errConflict,
 		},
 		{
 			name: "nested matching multierror",
-			err: errors.Wrap(terrors.MultiError([]error{
+			err: errors.Wrap(errors.Wrap(terrors.MultiError([]error{
 				storage.ErrOutOfOrderSample,
 				errors.New("foo"),
 				errors.New("bar"),
+			}), "baz"), "qux"),
+			threshold: 1,
+			exp:       errConflict,
+		},
+		{
+			name: "deep nested matching multierror",
+			err: errors.Wrap(terrors.MultiError([]error{
+				terrors.MultiError([]error{
+					errors.New("qux"),
+					status.Error(codes.AlreadyExists, "conflict"),
+					status.Error(codes.AlreadyExists, "conflict"),
+				}),
+				errors.New("foo"),
+				errors.New("bar"),
 			}), "baz"),
-			f:   isConflict,
-			out: 0,
+			threshold: 1,
+			exp:       errors.New("unexpected error: baz: 3 errors: 3 errors: qux; rpc error: code = AlreadyExists desc = conflict; rpc error: code = AlreadyExists desc = conflict; foo; bar"),
+		},
+	} {
+		got := cause(tc.err, tc.threshold)
+		if got == nil {
+			if tc.exp != nil {
+				t.Errorf("case %q: got nil, expected: %v", tc.name, tc.exp)
+				continue
+			}
+			continue
+		}
+		if got.Error() != tc.exp.Error() {
+			t.Errorf("case %q: got: %v, expected: %v", tc.name, got, tc.exp)
+			continue
+		}
+	}
+}
+
+func TestRootCause(t *testing.T) {
+	for _, tc := range []struct {
+		name string
+		err  error
+		exp  error
+	}{
+		{
+			name: "nested matching multierror",
+			err: errors.Wrap(errors.Wrap(terrors.MultiError([]error{
+				tsdb.ErrNotReady,
+				errors.New("foo"),
+				errors.New("bar"),
+			}), "baz"), "qux"),
+			exp: tsdb.ErrNotReady,
 		},
 		{
 			name: "deep nested matching multierror",
 			err: errors.Wrap(terrors.MultiError([]error{
 				terrors.MultiError([]error{
 					errors.New("qux"),
-					errors.New(strconv.Itoa(http.StatusConflict)),
+					status.Error(codes.AlreadyExists, "conflict"),
 				}),
 				errors.New("foo"),
 				errors.New("bar"),
 			}), "baz"),
-			f:   isConflict,
-			out: 0,
+			exp: errors.New("3 errors: 2 errors: qux; rpc error: code = AlreadyExists desc = conflict; foo; bar"),
 		},
 	} {
-		if n := countCause(tc.err, tc.f); n != tc.out {
-			t.Errorf("test case %s: expected %d, got %d", tc.name, tc.out, n)
+		got := rootCause(tc.err)
```

Comment:
What about old good `testutil.Equals`?

Reply:
`testutil.Equals` won't help in this case. We are not using sentinel or predefined errors. It never returns true.

Reply:
Well usually we do
```
if tc.exp != nil {
     testutil.NotOk(...
     testutil.Equals(t, tc.exp.Errors(),...
     continue
}
testutil.Ok(...
```